clients/src/main/java/org/apache/kafka/common/network/SSLTransportLayer.java

```diff
@@ -245,20 +248,23 @@ public void handshake() throws IOException {
                 case NEED_UNWRAP:
                     log.trace("SSLHandshake NEED_UNWRAP channelId {}, appReadBuffer pos {}, netReadBuffer pos {}, netWriteBuffer pos {}",
                               channelId, appReadBuffer.position(), netReadBuffer.position(), netWriteBuffer.position());
-                    handshakeResult = handshakeUnwrap(read);
+                    do {
+                        handshakeResult = handshakeUnwrap(read);
+                        if (handshakeResult.getStatus() == Status.BUFFER_OVERFLOW) {
+                            int currentAppBufferSize = applicationBufferSize();
+                            appReadBuffer = Utils.ensureCapacity(appReadBuffer, currentAppBufferSize);
+                            if (appReadBuffer.position() > currentAppBufferSize) {
+                                throw new IllegalStateException("Buffer underflow when available data size (" + appReadBuffer.position() +
+                                                                ") > packet buffer size (" + currentAppBufferSize + ")");
+                            }
+                        }
+                    } while (handshakeResult.getStatus() == Status.BUFFER_OVERFLOW);
```

Comment:
`SSLTransportLayerTest.testApplicationBufferResize()` does trigger BUFFER_OVERFLOW in unwrap() during handshake. The test does start with a very small buffer and the overflow is unlikely to be triggered when the initial buffer size is larger. Rather than make an assumption about buffer size, it feels neater to simply handle the overflow status. I am testing with IBM JRE.